servo/connectors/olas/prometheus.py

```diff
@@ -0,0 +1,145 @@
+
+        x = pn / 100.0 * buckets[-1][1]
+        # print(x, buckets[-1][1])
+        while low < high:
+            mid = (low + high) // 2
+            if buckets[mid][1] < x:
+                # print("mid < x", buckets[mid][1], x, "low", low, ">", mid +1)
+                low = mid + 1
+            else:
+                # print("mid >= x", buckets[mid][1], x, "high", high, ">", mid)
+                high = mid
+        if low == 0:
+            # the histogram has implicit (0,0) starting point
+            left = 0
+            low_v = 0
+        else:
+            left = float(buckets[low - 1][0])
+            low_v = buckets[low - 1][1]
+
+        high_v = buckets[low][1]
+        if low_v == high_v:
+            return left
+
+        right = float(buckets[low][0])
+        # print (buckets)
+        # print (low_v, x, high_v)
+        # print (left, right - left, x, x -low, high_v, low_v, high_v - low_v)
+        return left + (right - left) * (x - low_v) / (high_v - low_v)
+
+    def __str__(self):
+        buckets = [int(value) for label, value in self.buckets]
+        buckets = [n - c for c, n in zip(buckets[:-1], buckets[1:])]
+        while buckets and not buckets[-1]:
+            del buckets[-1]
+        dist = "\t".join(map(str, buckets))
+        return f'{sum(buckets)/60:.1f}\t[{dist}]'
+
+
+class PrometheusQuery:
+    def __init__(self, url):
+        self.url = url
+
+    async def querys(self, querys, timeout=10):
+        async def fetch(url, params={}, timeout=timeout):
+            async with aiohttp.ClientSession() as session:
+                async with session.get(url, params=params, timeout=timeout) as response:
+                    # servo.logger.info(f'PROM response {response}')
+                    if response.status == 200:
+                        obj = await response.json()
+                        return obj
+
+                    servo.logger.error(f'Promtheus status code "{response.status}"')
+            return {}
+
+        tasks = [asyncio.create_task(fetch(f'{self.url}/api/v1/{method}', params))
+                 for method, params in querys]
+        responses = await asyncio.gather(*tasks)
+        return responses
+
+    def extract_vector_list(self, res):
+        return [self.extract_vector(r) for r in res]
+
+    def extract_vector(self, r):
+        try:
+            reply = PrometheusReply.parse_obj(r)
+            result = reply.data.result
+            resultType = reply.data.resultType
+            if resultType == 'vector':
+                if not result:
+                    return 0
+                metric = PrometheusVector.parse_obj(reply.data.result[0])
+                return float(metric.value[-1])
+            elif resultType == 'matrix':
+                # More than one vector value. Leave to the caller to process it.
+                return result
+            else:
+                servo.logger.error(f'Unhandled promtheus result type "{resultType}"')
+        except Exception:
```

Comment:
We can't be doing blanket exception handlers like this. If you don't know what could go wrong you cannot just suppress it, return an arbitrary value, and hope for this best.